pkg/disk/hostapi/hostapi.go

```diff
@@ -271,90 +323,98 @@ func (imp DiskAPI) GetDiskNumberAndPage83ID(path string) (uint32, string, error)
-	disks := []Disk{}
-	err = json.Unmarshal([]byte(outString), &disks)
-	if err != nil {
-		return nil, err
-	}
-
 	m := make(map[uint32]DiskIDs)
-
-	for i := range disks {
-		diskNumber, page83, err := imp.GetDiskNumberAndPage83ID(disks[i].Path)
+	err := cim.WithCOMThread(func() error {
+		disks, err := cim.ListDisks(cim.DiskSelectorListForPathAndSerialNumber)
 		if err != nil {
-			return nil, err
+			return err
 		}
 
-		m[diskNumber] = DiskIDs{
-			Page83:       page83,
-			SerialNumber: disks[i].SerialNumber,
+		for _, disk := range disks {
+			path, err := cim.GetDiskPath(disk)
+			if err != nil {
+				return fmt.Errorf("failed to query disk path: %v, %w", disk, err)
+			}
+
+			sn, err := cim.GetDiskSerialNumber(disk)
+			if err != nil {
+				return fmt.Errorf("failed to query disk serial number: %v, %w", disk, err)
+			}
+
+			diskNumber, page83, err := imp.GetDiskNumberAndPage83ID(path)
+			if err != nil {
+				return err
+			}
+
+			m[diskNumber] = DiskIDs{
+				Page83:       page83,
+				SerialNumber: sn,
+			}
 		}
-	}
 
-	return m, nil
+		return nil
+	})
+	return m, err
 }
 
 func (imp DiskAPI) GetDiskStats(diskNumber uint32) (int64, error) {
-	cmd := fmt.Sprintf("(Get-Disk -Number %d).Size", diskNumber)
-	out, err := utils.RunPowershellCmd(cmd)
-	if err != nil || len(out) == 0 {
-		return -1, fmt.Errorf("error getting size of disk. cmd: %s, output: %s, error: %v", cmd, string(out), err)
-	}
-
-	reg, err := regexp.Compile("[^0-9]+")
-	if err != nil {
-		return -1, fmt.Errorf("error compiling regex. err: %v", err)
-	}
-	diskSizeOutput := reg.ReplaceAllString(string(out), "")
-
-	diskSize, err := strconv.ParseInt(diskSizeOutput, 10, 64)
+	// TODO: change to uint64 as it does not make sense to use int64 for size
+	size := int64(-1)
+	err := cim.WithCOMThread(func() error {
+		disk, err := cim.QueryDiskByNumber(diskNumber, cim.DiskSelectorListForSize)
+		if err != nil {
+			return err
+		}
 
-	if err != nil {
-		return -1, fmt.Errorf("error parsing size of disk. cmd: %s, output: %s, error: %v", cmd, diskSizeOutput, err)
-	}
+		size, err = cim.GetDiskSize(disk)
+		if err != nil {
+			return fmt.Errorf("failed to query size of disk %d. %v", diskNumber, err)
+		}
 
-	return diskSize, nil
+		return nil
+	})
+	return size, err
 }
 
 func (imp DiskAPI) SetDiskState(diskNumber uint32, isOnline bool) error {
-	cmd := fmt.Sprintf("(Get-Disk -Number %d) | Set-Disk -IsOffline $%t", diskNumber, !isOnline)
-	out, err := utils.RunPowershellCmd(cmd)
-	if err != nil {
-		return fmt.Errorf("error setting disk attach state. cmd: %s, output: %s, error: %v", cmd, string(out), err)
-	}
+	return cim.WithCOMThread(func() error {
+		disk, err := cim.QueryDiskByNumber(diskNumber, cim.DiskSelectorListForIsOffline)
+		if err != nil {
+			return err
+		}
+
+		isOffline, err := cim.IsDiskOffline(disk)
+		if err != nil {
+			return fmt.Errorf("error setting disk %d attach state. error: %v", diskNumber, err)
+		}
+
+		if isOnline == !isOffline {
```

Comment:
I think it'd be useful to have a log that says that the disk was already in the desired state.

Reply:
Sure.